File: scripts/ships/overlay.ts
/** `ship:overlay <id>`: our published GLB against a cached `ship:reference`, from the same camera.
 * Ours-only pixels are red, reference-only blue, shared grey; IoU per shot goes to summary.json.
 * A GameModels3D model is not centred on our midships, and its waterline datum can differ from ours, so the
 * reference is shifted: matched fore and aft on the waterline half-breadths, then refined fore and aft and up
 * and down on the side and top silhouettes. `--offset` overrides that. `--textured` renders the textured
 * reference and our model from the same cameras as paired images; `--sections` cuts both by the same planes.
 * Output: `.build/ships/<id>/overlay/`.
 * Diagnostic only: reference geometry is read here and never reaches `assets/` or `public/`. */
import { existsSync } from 'node:fs';
import { mkdir, readFile, writeFile } from 'node:fs/promises';
import { join, resolve } from 'node:path';
import { runBlender } from '../build/blender';
import { readReferenceMeta, referenceDirectory, selectTriangles } from '../construction/reference';
import { suggestedVehicles } from '../../tools/ship-overlay/reference';
import { cameraPin, parseCameraPose, parseVec3, type CameraPose } from '../browser/cameraPoses';
import { compareCuts, cut, drawSections, parseSections, type SectionSpec } from './sections';
import type { ShipDefinition, Vec3 } from '../../src/ships/blueprint';

const ROOT = resolve(import.meta.dir, '../..');
export const VIEWS = ['side', 'top', 'front', 'stern'] as const;
export type View = (typeof VIEWS)[number];
export const NAMED_SHOTS = [...VIEWS, 'bridge', 'aft'] as const;
type Box = [number | null, number | null, number | null, number | null, number | null, number | null];
export type Shot = { name: string; view: View; box: Box; px: number; margin?: number } | { name: string; pin: { eye: Vec3; target: Vec3; fov?: number; ortho?: number }; px: number };

export interface OverlayOptions {
  id: string;
  reference?: string;
  shots: string[];
  box?: number[];
  camera?: { pose: CameraPose; fov?: number; ortho?: number };
  parts?: string[];
  offset?: Vec3;
  align: 'silhouette' | 'fore-aft' | 'waterline' | 'none';
  px: number;
  fetch: boolean;
  textured: boolean;
  paint?: string;
  sections: SectionSpec[];
}

const HELP = `bun run ship:overlay <ship-id> [options]
  --reference <name|id>   cached ship:reference (default: the ship's suggested GameModels3D vehicle)
  --shots a,b,…            ${NAMED_SHOTS.join(', ')} (default side,top,front,bridge,aft)
  --box x0,y0,z0,x1,y1,z1  add side/top/front views of this box, clipped to it in depth
  --camera <preset|az,el[,m]> | --eye x,y,z --target x,y,z   one extra shot with this camera
  --fov <deg> | --ortho <m>  lens for that shot (perspective by default)
  --parts a,b,…            reference groups or part keys to draw (as ship:slice)
  --offset z|x,y,z         reference shift in metres instead of the fitted alignment
  --align silhouette|fore-aft|waterline|none   waterline match, then silhouette refinement fore and aft and up
                           and down (default); the same without the vertical; waterline only; none
  --textured               also render the textured reference and our model from each shot's camera under one
                           light: <shot>-reference.png, -ours.png and -pair.png in overlay/textured/
  --paint <id>             reference paint scheme for --textured (default: the source's default)
  --sections z=-40,y=6.5,x=0   cut both models by these planes (stations, plans, profiles) into
                           section-<axis><value>.png, reference blue and ours red on a metre grid
  --px <n>                 image width for full-length views (default 2400)
  --no-fetch               fail instead of running ship:reference for an uncached reference`;

export function parseOverlayArgs(argv: string[]): OverlayOptions {
  const value = (flag: string) => {
    const i = argv.indexOf(flag);
    if (i < 0) return undefined;
    const next = argv[i + 1];
    if (next === undefined || next.startsWith('--')) throw new Error(`${flag} needs a value.`);
    return next;
  };
  const switches = new Set(['--no-fetch', '--textured', '--help']);
  const id = argv.find((a, i) => !a.startsWith('--') && !(argv[i - 1]?.startsWith('--') && !switches.has(argv[i - 1]))) ?? argv[0];
  if (!id || id.startsWith('--')) throw new Error(HELP);
  const known = new Set(['--reference', '--shots', '--box', '--camera', '--eye', '--target', '--fov', '--ortho', '--parts', '--offset', '--align', '--px', '--no-fetch', '--textured', '--paint', '--sections', '--help']);
  for (const a of argv) if (a.startsWith('--') && !known.has(a)) throw new Error(`Unknown flag ${a}.\n${HELP}`);
  const shots = (value('--shots') ?? 'side,top,front,bridge,aft').split(',').map(s => s.trim()).filter(Boolean);
  for (const s of shots) if (!(NAMED_SHOTS as readonly string[]).includes(s)) throw new Error(`Unknown shot "${s}". Shots: ${NAMED_SHOTS.join(', ')}.`);
  const numbers = (flag: string, count: number[]) => {
    const text = value(flag);
    if (text === undefined) return undefined;
    const n = text.split(',').map(Number);
    if (!count.includes(n.length) || n.some(v => !Number.isFinite(v))) throw new Error(`${flag} takes ${count.join(' or ')} numbers; got "${text}".`);
    return n;
  };
  const box = numbers('--box', [6]);
  const offset = numbers('--offset', [1, 3]);
  const eye = value('--eye'), target = value('--target'), cameraText = value('--camera');
  if (eye && cameraText) throw new Error('Give either --camera or --eye/--target, not both.');
  const fov = numbers('--fov', [1])?.[0], ortho = numbers('--ortho', [1])?.[0];
  let camera: OverlayOptions['camera'];
  if (eye) camera = { pose: { eye: parseVec3(eye, '--eye'), ...(target ? { target: parseVec3(target, '--target') } : {}) }, fov, ortho };
  else if (cameraText) camera = { pose: parseCameraPose(cameraText), fov, ortho };
  else if (target || fov !== undefined || ortho !== undefined) throw new Error('--target, --fov and --ortho need --camera or --eye.');
  const align = value('--align') ?? 'silhouette';
  if (align !== 'silhouette' && align !== 'fore-aft' && align !== 'waterline' && align !== 'none') throw new Error('--align takes silhouette, fore-aft, waterline or none.');
  const sections = value('--sections');
  const px = numbers('--px', [1])?.[0] ?? 2400;
  if (px < 256 || px > 6000) throw new Error('--px takes 256 to 6000.');
  return {
    id, reference: value('--reference'), shots, box, camera, align, px, fetch: !argv.includes('--no-fetch'),
    textured: argv.includes('--textured'), paint: value('--paint'), sections: sections ? parseSections(sections) : [],
    parts: value('--parts')?.split(',').filter(Boolean),
    offset: offset ? (offset.length === 1 ? [0, 0, offset[0]] : offset) as Vec3 : undefined,
  };
}

/** Concrete shots in the runtime frame. Zooms centre on the definition's bridge viewpoint and on its
 * tallest structure abaft midships; boxes left open (null) are filled from the drawn geometry. */
export function resolveShots(options: OverlayOptions, definition: ShipDefinition): Shot[] {
  const length = definition.hull.length;
  const zoom = Math.min(56, length * .4);
  const structures = definition.structures ?? [];
  const tallestAft = structures
    .map(s => ({ z: s.footprint.reduce((n, p) => n + p[1], 0) / s.footprint.length, top: s.baseY + s.height }))
    .filter(s => s.z > length * .05)
    .sort((a, b) => b.top - a.top)[0]?.z ?? length * .22;
  const bridge = definition.viewpoints?.bridge?.[2] ?? -length * .15;
  const centred = (z: number): Box => [null, null, z - zoom / 2, null, null, z + zoom / 2];
  const shots: Shot[] = options.shots.map(name => {
    if (name === 'bridge') return { name, view: 'side' as View, box: centred(bridge), px: 1600, margin: 0 };
    if (name === 'aft') return { name, view: 'side' as View, box: centred(tallestAft), px: 1600, margin: 0 };
    return { name, view: name as View, box: [null, null, null, null, null, null] as Box, px: name === 'side' || name === 'top' ? options.px : 1200 };
  });
  if (options.box) {
    const b = options.box as Box;
    for (const view of ['side', 'top', 'front'] as View[]) shots.push({ name: `box-${view}`, view, box: b, px: 1600, margin: 0 });
  }
  if (options.camera) {
    const pin = cameraPin(options.camera.pose, length);
    shots.push({ name: 'camera', pin: { eye: pin.eye as Vec3, target: pin.target as Vec3, ...(options.camera.ortho ? { ortho: options.camera.ortho } : { fov: options.camera.fov ?? pin.fov ?? 40 }) }, px: 1600 });
  }
  return shots;
}

export async function overlay(options: OverlayOptions, root = ROOT) {
  const glb = join(root, 'public/models', options.id + '.glb');
  const definitionPath = join(root, 'public/models', options.id + '.json');
  if (!existsSync(glb) || !existsSync(definitionPath)) throw new Error(`No published model for "${options.id}". Build it with bun run ship:build ${options.id}.`);
  const definition = JSON.parse(await readFile(definitionPath, 'utf8')) as ShipDefinition;
  const reference = options.reference ?? suggestedVehicles[options.id];
  if (!reference) throw new Error(`No suggested reference for "${options.id}"; pass --reference <name|GameModels3D id>.`);
  if (!existsSync(join(referenceDirectory(root, reference), 'reference.json'))) {
    if (!options.fetch) throw new Error(`No cached reference "${reference}". Run bun run ship:reference ${reference}.`);
    console.error(`Caching reference ${reference} with ship:reference (network)…`);
    const child = Bun.spawnSync(['bun', 'scripts/construction/cli.ts', 'reference', reference], { cwd: root, stdout: 'ignore', stderr: 'inherit' });
    if (child.exitCode !== 0) throw new Error(`ship:reference ${reference} failed.`);
  }
  const meta = await readReferenceMeta(root, reference);
  const out = join(root, '.build/ships', options.id, 'overlay');
  await mkdir(out, { recursive: true });
  const hullGroup = meta.parts.some(p => p.group === 'hull') ? ['hull'] : undefined;
  const job = {
    glb, out,
    referenceMesh: join(referenceDirectory(root, reference), 'mesh.bin'),
    ranges: options.parts?.length ? selectTriangles(meta, options.parts) : [],
    hullRanges: hullGroup ? selectTriangles(meta, hullGroup) : [],
    align: options.align, offset: options.offset ?? null,
    shots: resolveShots(options, definition),
  };
  const jobFile = join(out, 'job.json');
  await writeFile(jobFile, JSON.stringify(job, null, 2) + '\n');
  const run = await runBlender(join(root, 'scripts/ships/overlay.py'), {}, { cwd: root, args: [jobFile], log: join(out, 'blender.log') });
  const line = run.stdout.split('\n').find(l => l.startsWith('OVERLAY_SUMMARY '));
  if (!line) throw new Error('Overlay render produced no summary; see ' + join(out, 'blender.log'));
  const summary = JSON.parse(line.slice('OVERLAY_SUMMARY '.length)) as OverlaySummary;
  const offset = summary.offset;
  const result: Record<string, unknown> = { ship: options.id, reference, ...summary, offsetFlag: '--offset ' + offset.join(','), hint: alignmentHint(summary, options) };
  if (options.sections.length) result.sections = await sectionCuts(root, glb, reference, offset, options, out);
  if (options.textured) {
    const { renderReference } = await import('../construction/referenceRender');
    const cameras = summary.shots.map(shot => ({ name: shot.name, ...shot.camera }));
    const rendered = await renderReference(root, meta, { cameras, model: glb, offset, paint: options.paint, parts: options.parts, out: join(out, 'textured') });
    result.textured = rendered.shots.map(shot => shot.file);
  }
  await writeFile(join(out, 'summary.json'), JSON.stringify(result, null, 2) + '\n');
  return result;
}

interface OverlaySummary {
  offset: Vec3;
  fitted: boolean;
  shots: { name: string; view: string; iou: number; camera: { eye: Vec3; target: Vec3; up: Vec3; ortho?: number; fov?: number; size: [number, number]; clip?: [number, number] } }[];
}

/** A word when the side silhouettes still disagree: usually a vertical datum the fit was not allowed to find. */
export function alignmentHint(summary: Pick<OverlaySummary, 'offset' | 'fitted' | 'shots'>, options: Pick<OverlayOptions, 'align'>): string | undefined {
  const side = summary.shots.find(s => s.name === 'side');
  const [, dy, dz] = summary.offset;
  if (summary.fitted && options.align === 'silhouette')
    return Math.abs(dy) >= 0.05
      ? `The reference is drawn ${Math.abs(dy).toFixed(2)} m ${dy > 0 ? 'lower' : 'higher'} than ours (a different waterline datum); ` +
        `pass --offset ${summary.offset.join(',')} to ship:reference --render and other comparisons.`
      : undefined;
  if (side && side.iou < 0.93) return `Side IoU ${side.iou} is low. Run without --offset and with the default --align silhouette to fit the vertical datum as well (fore-and-aft shift now ${dz}).`;
  return undefined;
}

/** `--sections`: the reference (shifted by the overlay offset) and our GLB cut by each plane. */
async function sectionCuts(root: string, glb: string, reference: string, offset: Vec3, options: OverlayOptions, out: string) {
  const { readReference } = await import('../construction/reference');
  const { parseGlb } = await import('../construction/referenceFile');
  const { meshView } = await import('../construction/slice');
  const mesh = await readReference(root, reference);
  const referenceView = meshView(mesh, options.parts);
  const parts = parseGlb(await readFile(glb));
  const positions = new Float32Array(parts.reduce((n, p) => n + p.positions.length, 0));
  const index = new Uint32Array(parts.reduce((n, p) => n + p.index.length, 0));
  let vertices = 0, at = 0;
  for (const part of parts) {
    positions.set(part.positions, vertices * 3);
    for (const i of part.index) index[at++] = i + vertices;
    vertices += part.positions.length / 3;
  }
  const ours = { positions, index, ranges: [{ first: 0, count: index.length / 3 }] };
  const box = options.box && { min: options.box.slice(0, 3) as Vec3, max: options.box.slice(3) as Vec3 };
  const rows = [];
  for (const spec of options.sections) {
    const a = cut(referenceView, spec, offset, box), b = cut(ours, spec, [0, 0, 0], box);
    const drawing = drawSections(spec, a, b);
    const file = join(out, `section-${spec.axis}${spec.value}.png`);
    await writeFile(file, drawing.png);
    rows.push({ plane: `${spec.axis}=${spec.value}`, file, right: drawing.right, up: drawing.up, pixelsPerMetre: drawing.pixelsPerMetre, segments: { reference: a.length, ours: b.length }, halfBreadth: compareCuts(spec, a, b) });
  }
  return rows;
}

if (import.meta.main) {
  try {
    const argv = process.argv.slice(2);
    if (!argv.length || argv.includes('--help')) { console.log(HELP); process.exit(argv.length ? 0 : 1); }
    console.log(JSON.stringify(await overlay(parseOverlayArgs(argv)), null, 2));
  } catch (error) {
    console.error(error instanceof Error ? error.message : error);
    process.exit(1);
  }
}

File: scripts/ships/overlay.py
"""Silhouette overlay of a published ship GLB against a cached ship:reference.

Run by scripts/ships/overlay.ts in background Blender with one argument: a JSON job file.
Both models are drawn in the runtime frame (+X starboard, +Y up, -Z bow, waterline Y=0);
the reference is shifted by the job's offset, or by one measured from the waterlines and
refined on the side and top silhouettes (fore and aft, and up and down on the side view). Each shot renders ours and the
reference alone with the same camera and transparent film, then composites: ours only red,
reference only blue, both in the reference's own shading.
"""
import bpy
import json
import math
import os
import struct
import sys
import numpy as np
from mathutils import Matrix, Vector

job = json.loads(open(sys.argv[sys.argv.index('--') + 1]).read())
bpy.ops.wm.read_factory_settings(use_empty=True)
scene = bpy.context.scene

# Blender frame after the glTF import: (X, Y, Z) = (x, -z, y) in runtime terms.
def to_blender(p):
    return np.stack([p[:, 0], -p[:, 2], p[:, 1]], axis=1)


def to_runtime(b):
    return np.stack([b[:, 0], b[:, 2], -b[:, 1]], axis=1)


bpy.ops.import_scene.gltf(filepath=job['glb'])
ours = [o for o in scene.objects if o.type == 'MESH']
ours_root = bpy.data.collections.new('ours'); scene.collection.children.link(ours_root)
for o in ours:
    for c in list(o.users_collection): c.objects.unlink(o)
    ours_root.objects.link(o)


def world_vertices(objs):
    chunks = []
    for o in objs:
        me = o.data; n = len(me.vertices)
        if not n: continue
        co = np.empty(n * 3, np.float32); me.vertices.foreach_get('co', co)
        m = np.array(o.matrix_world, np.float64)
        v = co.reshape(-1, 3).astype(np.float64) @ m[:3, :3].T + m[:3, 3]
        chunks.append(v)
    return np.concatenate(chunks) if chunks else np.zeros((0, 3))


ours_rt = to_runtime(world_vertices(ours))


def world_triangles(objs):
    """Runtime-frame triangles (n, 3, 3) of the meshes, only those that reach the waterline band."""
    chunks = []
    for o in objs:
        me = o.data
        if not len(me.polygons): continue
        me.calc_loop_triangles()
        co = np.empty(len(me.vertices) * 3, np.float32); me.vertices.foreach_get('co', co)
        m = np.array(o.matrix_world, np.float64)
        v = to_runtime(co.reshape(-1, 3).astype(np.float64) @ m[:3, :3].T + m[:3, 3])
        tri = np.empty(len(me.loop_triangles) * 3, np.int32); me.loop_triangles.foreach_get('vertices', tri)
        t = v[tri.reshape(-1, 3)]
        keep = (t[:, :, 1].min(1) < .5) & (t[:, :, 1].max(1) > -.5)
        chunks.append(t[keep])
    return np.concatenate(chunks) if chunks else np.zeros((0, 3, 3))


def waterplane(tris, level=.0137):
    """Points where the triangles' edges cross a plane just above the waterline. Authored hulls
    often carry a vertex row exactly at y = 0, whose edges would never register as crossing it."""
    out = []
    for i, j in [(0, 1), (1, 2), (2, 0)]:
        a, b = tris[:, i], tris[:, j]
        cross = ((a[:, 1] - level) * (b[:, 1] - level)) < 0
        t = ((a[cross, 1] - level) / (a[cross, 1] - b[cross, 1]))[:, None]
        out.append(a[cross] + t * (b[cross] - a[cross]))
    return np.concatenate(out) if out else np.zeros((0, 3))


ours_water = waterplane(world_triangles(ours))

raw = open(job['referenceMesh'], 'rb').read()
nv, ni = struct.unpack_from('<II', raw, 8)
P = np.frombuffer(raw, np.float32, nv * 3, 16).reshape(-1, 3).astype(np.float64)
I = np.frombuffer(raw, np.uint32, ni, 16 + nv * 12).reshape(-1, 3)
tris = np.concatenate([I[r['first']:r['first'] + r['count']] for r in job['ranges']]) if job['ranges'] else I
hull_tris = np.concatenate([I[r['first']:r['first'] + r['count']] for r in job['hullRanges']]) if job['hullRanges'] else tris


def waterline_span(points):
    if len(points) < 8:
        return None
    lo, hi = np.percentile(points[:, 2], [.2, 99.8])
    return float(lo), float(hi)


def waterline_profile(points, grid):
    """Waterline half-breadth along the grid: the largest |x| per 0.5 m, interpolated where points exist."""
    edges = np.arange(grid[0], grid[-1] + .5, .5)
    k = np.clip(((points[:, 2] - edges[0]) / .5).astype(int), 0, len(edges) - 1)
    widest = np.full(len(edges), -1.0); np.maximum.at(widest, k, np.abs(points[:, 0]))
    have = widest >= 0
    if have.sum() < 2:
        return np.zeros(len(grid))
    centres = edges[have] + .25
    profile = np.interp(grid, centres, widest[have], left=0, right=0)
    profile[(grid < centres[0] - .25) | (grid > centres[-1] + .25)] = 0
    return profile


ref_water = waterplane(P[hull_tris])
ours_span = waterline_span(ours_water)
ref_span = waterline_span(ref_water)
if job['offset'] is not None:
    offset = job['offset']
elif job['align'] != 'none' and ours_span and ref_span:
    # Start from the span midpoints, then slide the reference's waterline half-breadths along ours
    # (±6 m in 5 cm steps) to the least absolute difference: exact when the hull shapes agree.
    guess = (ours_span[0] + ours_span[1]) / 2 - (ref_span[0] + ref_span[1]) / 2
    grid = np.arange(min(ours_span[0], ref_span[0] + guess) - 10, max(ours_span[1], ref_span[1] + guess) + 10, .05)
    mine = waterline_profile(ours_water, grid)
    best = min(((float(np.abs(mine - waterline_profile(ref_water + [0, 0, guess + shift], grid)).sum()), shift)
                for shift in np.arange(-6, 6.001, .05)))
    offset = [0.0, 0.0, guess + best[1]]
else:
    offset = [0.0, 0.0, 0.0]
Pr = P + np.array(offset)
used = np.unique(tris); remap = np.full(len(P), -1, np.int64); remap[used] = np.arange(len(used))
ref_mesh = bpy.data.meshes.new('reference')
ref_mesh.from_pydata(to_blender(Pr[used]).tolist(), [], remap[tris].tolist())
ref_mesh.update()
ref_obj = bpy.data.objects.new('reference', ref_mesh)
ref_root = bpy.data.collections.new('reference'); scene.collection.children.link(ref_root); ref_root.objects.link(ref_obj)
ref_rt = Pr[used]

scene.render.engine = 'BLENDER_WORKBENCH'
shading = scene.display.shading
shading.light = 'STUDIO'; shading.color_type = 'SINGLE'; shading.single_color = (.62, .64, .66)
shading.show_cavity = True; shading.cavity_type = 'BOTH'; shading.show_shadows = False
scene.render.film_transparent = True
scene.view_settings.view_transform = 'Standard'
camera_data = bpy.data.cameras.new('overlay'); camera_data.type = 'ORTHO'; camera_data.sensor_fit = 'HORIZONTAL'
camera = bpy.data.objects.new('overlay', camera_data); scene.collection.objects.link(camera); scene.camera = camera

# View: (looking direction, image right, image up), runtime axes. Bow is to the right in
# side and top views; front looks aft from ahead, stern looks forward from astern.
VIEWS = {
    'side': ((-1, 0, 0), (0, 0, -1), (0, 1, 0)),
    'top': ((0, -1, 0), (0, 0, -1), (-1, 0, 0)),
    'front': ((0, 0, 1), (-1, 0, 0), (0, 1, 0)),
    'stern': ((0, 0, -1), (1, 0, 0), (0, 1, 0)),
}
both = np.concatenate([ours_rt, ref_rt])


def bounds(box):
    lo = np.array([-np.inf if v is None else v for v in box[:3]], np.float64)
    hi = np.array([np.inf if v is None else v for v in box[3:]], np.float64)
    inside = np.all((both >= lo) & (both <= hi), axis=1)
    pts = both[inside] if inside.any() else both
    glo, ghi = pts.min(0), pts.max(0)
    lo = np.where(np.isfinite(lo), lo, glo); hi = np.where(np.isfinite(hi), hi, ghi)
    return lo, hi


def ortho(view, box, px, margin):
    """Aim the orthographic camera at the box from one of VIEWS; returns image and metre sizes."""
    camera_data.type = 'ORTHO'
    look, right, up = (np.array(v, np.float64) for v in VIEWS[view])
    lo, hi = bounds(box)
    lo = lo - margin; hi = hi + margin
    centre = (lo + hi) / 2
    span = hi - lo
    width = float(abs(span @ right)); height = float(abs(span @ up)); depth = float(abs(span @ look))
    px_w = px; px_h = max(64, min(6000, int(round(px_w * height / max(width, 1e-6)))))
    scene.render.resolution_x, scene.render.resolution_y = px_w, px_h
    camera_data.ortho_scale = width
    eye_rt = centre - look * (depth / 2 + .01)
    to_b = lambda v: Vector((v[0], -v[2], v[1]))
    r, u, f = to_b(right).normalized(), to_b(up).normalized(), to_b(look).normalized()
    camera.matrix_world = Matrix(((r.x, u.x, -f.x, 0), (r.y, u.y, -f.y, 0), (r.z, u.z, -f.z, 0), (0, 0, 0, 1)))
    camera.location = to_b(eye_rt)
    camera_data.clip_start = .001; camera_data.clip_end = depth + .02
    global last_camera
    last_camera = dict(eye=eye_rt.tolist(), target=(eye_rt + look).tolist(), up=up.tolist(), ortho=width,
                       size=[px_w, px_h], clip=[.001, depth + .02])
    return px_w, px_h, width, height


def render(name, collection_visible, path):
    ours_root.hide_render = collection_visible != 'ours'
    ref_root.hide_render = collection_visible != 'reference'
    scene.render.filepath = path
    bpy.ops.render.render(write_still=True)
    image = bpy.data.images.load(path)
    w, h = image.size
    px = np.empty(w * h * 4, np.float32); image.pixels.foreach_get(px)
    bpy.data.images.remove(image)
    return px.reshape(h, w, 4)


def aim(eye_rt, target_rt):
    """Point the camera from eye to target with runtime +Y up."""
    to_b = lambda v: Vector((v[0], -v[2], v[1]))
    eye, target = to_b(eye_rt), to_b(target_rt)
    camera.location = eye
    camera.rotation_euler = (target - eye).to_track_quat('-Z', 'Y').to_euler()


def overlap(a, b, reach_x, reach_y):
    """Intersection of mask a with mask b moved by every (dy, dx) pixel shift within reach, by FFT
    cross-correlation: inter[dy + reach_y, dx + reach_x] = sum a[p] * b[p - (dy, dx)]. The masks are
    rendered with at least `reach` of empty margin, so the circular shift equals the linear one."""
    fa = np.fft.rfft2(a.astype(np.float32))
    fb = np.fft.rfft2(b.astype(np.float32))
    c = np.fft.irfft2(fa * np.conj(fb), s=a.shape)
    rows = np.arange(-reach_y, reach_y + 1) % a.shape[0]
    cols = np.arange(-reach_x, reach_x + 1) % a.shape[1]
    return np.rint(c[np.ix_(rows, cols)])


def peak(values, k):
    """Sub-step position of a maximum from the parabola through it and its neighbours."""
    if 0 < k < len(values) - 1:
        l, c, r = values[k - 1], values[k], values[k + 1]
        if l - 2 * c + r < 0: return .5 * (l - r) / (l - 2 * c + r)
    return 0.0


def refine(offset, vertical):
    """Slide the reference's side and top silhouettes along ours (±4 m at about 4 cm a pixel) to the
    greatest combined overlap. Both views put the bow to the right, so a fore-and-aft shift is a
    horizontal image shift; with `vertical` the side view also slides up and down (±3 m), which finds
    a reference whose waterline datum differs from ours (Bismarck's sits 0.85 m low). The waterline
    match alone left Iowa 0.4 m short of the best fit."""
    reach, reach_up = 4.0, 3.0 if vertical else 0.0
    px_w = min(6000, int(math.ceil((np.ptp(both[:, 2]) + 2 * reach) / .04)))
    inter = union_parts = None
    for view in ('side', 'top'):
        _, _, width, _ = ortho(view, [None] * 6, px_w, reach)
        a = render('align', 'ours', job['out'] + '/align-ours.png')[:, :, 3] > .5
        b = render('align', 'reference', job['out'] + '/align-reference.png')[:, :, 3] > .5
        metres = width / px_w
        steps = int(reach / metres)
        rows = int(reach_up / metres) if view == 'side' else 0
        counts = overlap(a, b, steps, rows)
        if view == 'top':
            counts = counts[:1]
        area = float(a.sum() + b.sum())
        inter = counts if inter is None else inter + counts
        union_parts = area if union_parts is None else union_parts + area
        side_rows = rows if view == 'side' else side_rows
    for side in ('ours', 'reference'): os.remove(job['out'] + f'/align-{side}.png')
    iou = inter / np.maximum(union_parts - inter, 1)
    ky, kx = np.unravel_index(int(np.argmax(iou)), iou.shape)
    fx = peak(iou[ky], kx)
    fy = peak(iou[:, kx], ky)
    # Image right is runtime -Z, so moving the reference right by s pixels is a shift of -s * metres.
    # Rows count up from the bottom of the image, which is runtime +Y.
    dz = -(kx - steps + fx) * metres
    dy = (ky - side_rows + fy) * metres if vertical else 0.0
    return [offset[0], offset[1] + dy, offset[2] + dz]


if job['offset'] is None and job['align'] in ('silhouette', 'fore-aft'):
    start = list(offset)
    offset = refine(offset, job['align'] == 'silhouette')
    shift_y, shift_z = offset[1] - start[1], offset[2] - start[2]
    ref_obj.location = (0, -shift_z, shift_y)
    ref_rt = ref_rt + [0, shift_y, shift_z]
    both = np.concatenate([ours_rt, ref_rt])

summary = []
for shot in job['shots']:
    if 'pin' in shot:
        pin = shot['pin']; px_w = shot['px']; px_h = int(round(px_w * .625))
        scene.render.resolution_x, scene.render.resolution_y = px_w, px_h
        if pin.get('ortho'):
            camera_data.type = 'ORTHO'; camera_data.ortho_scale = pin['ortho']
        else:
            camera_data.type = 'PERSP'; camera_data.sensor_width = 36
            camera_data.lens = 18 / math.tan(math.radians(pin.get('fov', 40)) / 2)
        camera_data.clip_start = .05; camera_data.clip_end = 5000
        aim(np.array(pin['eye'], np.float64), np.array(pin['target'], np.float64))
        width = height = 0.0
        # The same camera as reference_render.py describes it: a vertical field of view.
        fov = pin.get('fov', 40)
        lens = {'ortho': pin['ortho']} if pin.get('ortho') else {'fov': math.degrees(2 * math.atan(math.tan(math.radians(fov) / 2) * px_h / px_w))}
        last_camera = dict(eye=list(pin['eye']), target=list(pin['target']), up=[0, 1, 0], size=[px_w, px_h], **lens)
    else:
        px_w, px_h, width, height = ortho(shot['view'], shot['box'], shot['px'], shot.get('margin', 2.0))
    base = job['out'] + '/' + shot['name']
    a = render(shot['name'], 'ours', base + '-ours.png')
    b = render(shot['name'], 'reference', base + '-reference.png')
    ma = a[:, :, 3] > .5; mb = b[:, :, 3] > .5
    out = np.ones((px_h, px_w, 4), np.float32); out[:, :, :3] = .92
    shade = b[:, :, :3].mean(axis=2, keepdims=True) * .8
    out[ma & mb, :3] = np.repeat(shade, 3, axis=2)[ma & mb]
    out[ma & ~mb, :3] = (.86, .16, .16)
    out[mb & ~ma, :3] = (.16, .35, .86)
    image = bpy.data.images.new(shot['name'] + '-overlay', px_w, px_h, alpha=True)
    image.pixels.foreach_set(out.ravel())
    image.filepath_raw = base + '.png'; image.file_format = 'PNG'; image.save()
    union = int((ma | mb).sum())
    summary.append(dict(name=shot['name'], view=shot.get('view', 'camera'), file=base + '.png',
                        widthM=round(width, 2), heightM=round(height, 2), pixels=[px_w, px_h],
                        iou=round(float((ma & mb).sum()) / max(union, 1), 4),
                        oursOnly=round(float((ma & ~mb).sum()) / max(union, 1), 4),
                        referenceOnly=round(float((mb & ~ma).sum()) / max(union, 1), 4),
                        camera=last_camera))

result = dict(offset=[round(v, 4) for v in offset], fitted=job['offset'] is None and job['align'] != 'none',
              waterline=dict(ours=ours_span and [round(v, 3) for v in ours_span], reference=ref_span and [round(v, 3) for v in ref_span]),
              shots=summary)
open(job['out'] + '/summary.json', 'w').write(json.dumps(result, indent=2) + '\n')
print('OVERLAY_SUMMARY ' + json.dumps(result))
